### PR TITLE
feat: add fetch feature

### PR DESCRIPTION
**Pull request by Doozers**
Add a `fetch` button to fetch the current state of the repo.
It created a new param on the get: `/graph` request. (withFetch=boolean)

This only work without http caching. (because it will consider `fetch` and and `generate` as the same action. [I guess])

demo here :
https://www.loom.com/share/8e0a1f9d50264cb89f2bbb0d72ffb8d5

┆Issue is synchronized with this [Trello card](https://trello.com/c/fQLQA7AX)

### CODE REVIEW

File: internal/dvserver/server.go [hunk 1]
```diff
@@ -173,6 +173,15 @@ func New(ctx context.Context, h *cayley.Handle, schema *schema.Config, opts Opts
 				OrigName:     true,
 			}),
 			runtime.WithProtoErrorHandler(runtime.DefaultHTTPProtoErrorHandler),
+			runtime.WithIncomingHeaderMatcher(func(key string) (string, bool) {
+				switch key {
+				case "Authorization":
+					fmt.Println(key)
```
Comment (moul) [inline internal/dvserver/server.go:179]:
```suggestion
```

eventually keep it in a Debug logger.

File: internal/dvserver/api.go [hunk 2]
```diff
@@ -2,18 +2,48 @@ package dvserver
...
 	"moul.io/depviz/v3/internal/dvstore"
 )
 
+func getToken(ctx context.Context) (string, error) {
+	md, _ := metadata.FromIncomingContext(ctx)
+
+	var gitHubToken string
+	if md["authorization"] != nil {
+		// skip "Basic"
+		gitHubToken = md["authorization"][1][6:]
```
Comment (moul) [inline internal/dvserver/api.go:22]: can panic

File: internal/dvserver/api.go [hunk 3]
```diff
@@ -2,18 +2,48 @@ package dvserver
...
+		// prevent empty token (skip prefix)
+		if gitHubToken == base64.StdEncoding.EncodeToString([]byte("depviz:")) {
+			gitHubToken = ""
```
Comment (moul) [inline internal/dvserver/api.go:25]:
```suggestion
			return gitHubToken, nil
```

could probably be rewritten this way for improved readability.

File: internal/dvserver/api.go [hunk 4]
```diff
@@ -35,9 +66,22 @@ func (s *service) Graph(ctx context.Context, in *Graph_Input) (*Graph_Output, er
...
+		tasks, err = dvstore.LoadTasks(s.h, s.schema, filters, s.opts.Logger)
+		if err != nil {
+			return nil, fmt.Errorf("load tasks: %w", err)
+		}
+	} else {
+		tasks, err = dvstore.LoadTasks(s.h, s.schema, filters, s.opts.Logger)
+		if err != nil {
+			return nil, fmt.Errorf("load tasks: %w", err)
+		}
 	}
```
Comment (moul) [inline internal/dvserver/api.go:85]:
```suggestion
	if filters.WithFetch && gitHubToken != "" {
		_, err := dvcore.PullAndSave(filters.Targets, s.h, s.schema, gitHubToken, false, s.opts.Logger)
		if err != nil {
			return nil, fmt.Errorf("pull: %w", err)
		}
	}

	var tasks dvmodel.Tasks
	tasks, err = dvstore.LoadTasks(s.h, s.schema, filters, s.opts.Logger)
	if err != nil {
		return nil, fmt.Errorf("load tasks: %w", err)
	}
```

Comment (Doozers) [reply internal/dvserver/api.go:85]: yes obviously

File: internal/dvserver/api.go (hunk 3 above)
Comment (Doozers) [reply internal/dvserver/api.go:25]: 👍

File: internal/dvserver/api.go [hunk 5]
```diff
@@ -2,18 +2,48 @@ package dvserver
...
+	}
+	return gitHubToken, nil
+}
+
 func (s *service) Graph(ctx context.Context, in *Graph_Input) (*Graph_Output, error) {
 	s.opts.Logger.Debug("graph", zap.Any("in", in))
-	// validation
+
+	// retrieve token
+	gitHubToken, err := getToken(ctx)
```
[the remark this answers is not in the document]
Comment (Doozers) [reply internal/dvserver/api.go:42]: I moved it as it's old place but tell me if you want it somewhere else

**Comment by moul** [comment]
can we close this one?

**Comment by Doozers** [comment]
yes

**Pull request closed without merging**